File: frontend/src/lib/api/findings.ts
import { apiGet, apiPatch, apiPost } from './client';
import type { Page } from './chatbot';

export type FindingSeverity = 'low' | 'medium' | 'high' | 'critical';
export type FindingStatus =
  | 'open'
  | 'confirmed'
  | 'in_progress'
  | 'fixed'
  | 'verified'
  | 'closed'
  | 'false_positive'
  | 'accepted_risk'
  | 'reopened';

export interface Finding {
  id: string;
  project_id: string;
  scan_run_id: string | null;
  fingerprint: string;
  rule_id: string;
  category: string;
  title: string;
  evidence: string;
  impact: string;
  remediation: string;
  severity: FindingSeverity;
  status: FindingStatus;
  target: string;
  cve_id: string | null;
  assignee_user_id: string | null;
  deadline: string | null;
  /** Computed at read time (backend.services.sla.is_overdue) - never a
   * stored column. */
  is_overdue: boolean;
  verification_notes: string;
  resolution_reason: string | null;
  first_seen_scan_run_id: string | null;
  last_seen_scan_run_id: string | null;
  closed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface FindingTransition {
  id: string;
  finding_id: string;
  from_status: string;
  to_status: string;
  actor_user_id: string;
  reason: string | null;
  created_at: string;
}

export interface FindingCreateInput {
  rule_id: string;
  category: string;
  title: string;
  evidence?: string;
  impact?: string;
  remediation?: string;
  severity: FindingSeverity;
  target: string;
  cve_id?: string | null;
  assignee_user_id?: string | null;
}

export interface FindingFilters {
  page?: number;
  pageSize?: number;
  status?: FindingStatus;
  severity?: FindingSeverity;
  assigneeUserId?: string;
  overdue?: boolean;
}

/** A project member eligible to be a Finding's assignee (Task 4) - project
 * role developer/security/owner, never viewer. No local User table exists
 * in this app, so (same as ProjectMember) there is no display name/email -
 * just the raw user_id. */
export interface EligibleAssignee {
  user_id: string;
  project_role: 'developer' | 'security' | 'owner';
}

/** A cross-project "My Tasks" row: a Finding plus its parent project's name. */
export interface MyTask extends Finding {
  project_name: string;
}

export interface MyTaskFilters {
  page?: number;
  pageSize?: number;
  status?: FindingStatus;
  severity?: FindingSeverity;
  overdue?: boolean;
}

export function listFindings(projectId: string, filters: FindingFilters = {}): Promise<Page<Finding>> {
  const params = new URLSearchParams();
  params.set('page', String(filters.page ?? 1));
  params.set('page_size', String(filters.pageSize ?? 20));
  if (filters.status) params.set('status', filters.status);
  if (filters.severity) params.set('severity', filters.severity);
  if (filters.assigneeUserId) params.set('assignee_user_id', filters.assigneeUserId);
  if (filters.overdue !== undefined) params.set('overdue', String(filters.overdue));
  return apiGet<Page<Finding>>(`/api/projects/${projectId}/findings?${params.toString()}`);
}

export function getFinding(projectId: string, findingId: string): Promise<Finding> {
  return apiGet<Finding>(`/api/projects/${projectId}/findings/${findingId}`);
}

export function createFinding(projectId: string, input: FindingCreateInput): Promise<Finding> {
  return apiPost<Finding>(`/api/projects/${projectId}/findings`, input);
}

export function transitionFinding(
  projectId: string,
  findingId: string,
  toStatus: FindingStatus,
  reason?: string,
): Promise<Finding> {
  return apiPost<Finding>(`/api/projects/${projectId}/findings/${findingId}/transition`, {
    to_status: toStatus,
    reason: reason ?? null,
  });
}

export function setAssignee(
  projectId: string,
  findingId: string,
  assigneeUserId: string | null,
): Promise<Finding> {
  return apiPatch<Finding>(`/api/projects/${projectId}/findings/${findingId}/assignee`, {
    assignee_user_id: assigneeUserId,
  });
}

export function listEligibleAssignees(projectId: string): Promise<{ items: EligibleAssignee[] }> {
  return apiGet<{ items: EligibleAssignee[] }>(`/api/projects/${projectId}/findings/eligible-assignees`);
}

export function listMyTasks(filters: MyTaskFilters = {}): Promise<Page<MyTask>> {
  const params = new URLSearchParams();
  params.set('page', String(filters.page ?? 1));
  params.set('page_size', String(filters.pageSize ?? 20));
  if (filters.status) params.set('status', filters.status);
  if (filters.severity) params.set('severity', filters.severity);
  if (filters.overdue !== undefined) params.set('overdue', String(filters.overdue));
  return apiGet<Page<MyTask>>(`/api/findings/my-tasks?${params.toString()}`);
}

File: backend/schemas/findings.py
"""Request/response models for findings and finding transitions."""
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from backend.schemas.common import Page, UtcDatetime

FindingSeverity = Literal["low", "medium", "high", "critical"]
FindingStatus = Literal[
    "open",
    "confirmed",
    "in_progress",
    "fixed",
    "verified",
    "closed",
    "false_positive",
    "accepted_risk",
    "reopened",
]


class FindingCreate(BaseModel):
    rule_id: str = Field(min_length=1, max_length=100)
    category: str = Field(min_length=1, max_length=100)
    title: str = Field(min_length=1, max_length=240)
    evidence: str = Field(default="")
    impact: str = Field(default="")
    remediation: str = Field(default="")
    severity: FindingSeverity
    target: str = Field(min_length=1, max_length=2048)
    cve_id: Optional[str] = Field(default=None, max_length=32)
    assignee_user_id: Optional[UUID] = None


class FindingTransitionRequest(BaseModel):
    to_status: FindingStatus
    reason: Optional[str] = Field(default=None, max_length=4000)


class FindingAssigneeUpdate(BaseModel):
    assignee_user_id: Optional[UUID] = None


class FindingItem(BaseModel):
    id: UUID
    project_id: UUID
    scan_run_id: Optional[UUID] = None
    fingerprint: str
    rule_id: str
    category: str
    title: str
    evidence: str
    impact: str
    remediation: str
    severity: FindingSeverity
    status: FindingStatus
    target: str
    cve_id: Optional[str] = None
    assignee_user_id: Optional[UUID] = None
    deadline: Optional[UtcDatetime] = None
    #: Computed at read time (backend.services.sla.is_overdue) - never a
    #: stored column, so it can never drift out of sync with "now" or the
    #: finding's current status. See backend/api/findings.py's _finding_dict.
    is_overdue: bool = False
    verification_notes: str
    resolution_reason: Optional[str] = None
    first_seen_scan_run_id: Optional[UUID] = None
    last_seen_scan_run_id: Optional[UUID] = None
    closed_at: Optional[UtcDatetime] = None
    created_at: UtcDatetime
    updated_at: UtcDatetime


class FindingPage(Page[FindingItem]):
    pass


class EligibleAssigneeItem(BaseModel):
    """A project member eligible to be a Finding's assignee (Task 4) -
    developer/security/owner project roles, never viewer. No local ``User``
    table exists in this app (identity comes from Supabase ``auth.users``) -
    same as ``ProjectMemberItem`` (Task 1), this deliberately returns only
    ``user_id``/``project_role`` rather than inventing a display-name/email
    resolution the rest of the app doesn't have either."""

    user_id: UUID
    project_role: Literal["developer", "security", "owner"]


class EligibleAssigneeList(BaseModel):
    items: list[EligibleAssigneeItem]


class MyTaskItem(FindingItem):
    """A ``FindingItem`` plus its parent project's name, for the
    cross-project "My Tasks" view (Task 4)."""

    project_name: str


class MyTaskPage(Page[MyTaskItem]):
    pass


class FindingTransitionItem(BaseModel):
    id: UUID
    finding_id: UUID
    from_status: str
    to_status: str
    actor_user_id: UUID
    reason: Optional[str] = None
    created_at: UtcDatetime
